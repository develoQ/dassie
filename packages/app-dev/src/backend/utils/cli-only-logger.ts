import {
  createCliFormatter,
  createEnableChecker,
  createLoggerFactory,
} from "@dassie/lib-logger"

export const createCliOnlyLogger = createLoggerFactory({
  enableChecker: createEnableChecker(process.env["DEBUG"] ?? ""),
  formatter: createCliFormatter(),
})
