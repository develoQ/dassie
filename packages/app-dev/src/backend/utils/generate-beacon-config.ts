import { BEACONS_START_PORT } from "../constants/ports"

const ENTRYPOINT = new URL("../../runner/launchers/beacon", import.meta.url)
  .pathname
const LOCAL_PATH = new URL("../../../../../local", import.meta.url).pathname

export const beaconIndexToId = (index: number) => `b${index + 1}`
export const beaconIndexToPort = (index: number) => BEACONS_START_PORT + index

export const generateBeaconConfig = (index: number) => {
  const id = beaconIndexToId(index)
  const port = beaconIndexToPort(index)

  return {
    id,
    port,
    config: {
      host: `${id}.localhost`,
      port: beaconIndexToPort(index),
      tlsWebCertFile: `${LOCAL_PATH}/tls/${id}.localhost/web-${id}.localhost.pem`,
      tlsWebKeyFile: `${LOCAL_PATH}/tls/${id}.localhost/web-${id}.localhost-key.pem`,
    },
    url: `https://${id}.localhost:${port}/`,
    entry: ENTRYPOINT,
  }
}
