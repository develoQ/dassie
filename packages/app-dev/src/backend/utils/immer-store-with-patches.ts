import { Patch, enablePatches, produceWithPatches } from "immer"

import { Store, createStore } from "@xen-ilp/lib-reactive"

import type { ImmerProducer } from "./immer-store"

type PatchesTuple<TState> = readonly [
  nextState: TState,
  patches: readonly Patch[],
  inversePatches: readonly Patch[]
]

enablePatches()

export const createImmerStoreWithPatches = <TState>(
  initialValue: TState
): Store<PatchesTuple<TState>, ImmerProducer<TState>> => {
  const store = createStore<PatchesTuple<TState>>([
    initialValue,
    [],
    [],
  ] as const)

  const reducer =
    (producer: ImmerProducer<TState>) =>
    (previousState: PatchesTuple<TState>) => {
      return produceWithPatches(previousState[0], producer)
    }

  const emit = (producer: ImmerProducer<TState>) => {
    store.emit(reducer(producer))
  }

  return {
    ...store,
    emit,
  }
}
