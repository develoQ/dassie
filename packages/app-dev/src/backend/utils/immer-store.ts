import produce from "immer"
import type { WritableDraft } from "immer/dist/internal"

import { Store, createStore } from "@dassie/lib-reactive"

export type ImmerProducer<TState> = (
  previousState: WritableDraft<TState>
) => void

export type ImmerStore<TState> = Store<TState, ImmerProducer<TState>>

export const createImmerStore = <TState>(
  initialValue: TState
): ImmerStore<TState> => {
  const store = createStore<TState>(initialValue)
  const reducer =
    (producer: ImmerProducer<TState>) => (previousState: TState) => {
      return produce(previousState, producer)
    }

  const update = (producer: ImmerProducer<TState>) => {
    return store.update(reducer(producer))
  }

  return {
    ...store,
    update,
  }
}
