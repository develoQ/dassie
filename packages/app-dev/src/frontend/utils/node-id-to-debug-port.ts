export const nodeIdToDebugUrl = (nodeId: string) => {
  return `ws://localhost:${Number(nodeId.slice(1)) + 5000 - 1}`
}
