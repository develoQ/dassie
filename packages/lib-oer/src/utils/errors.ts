export class ParseError extends Error {
  constructor(
    message: string,
    readonly data: Uint8Array,
    readonly offset: number
  ) {
    super(message)

    this.name = "ParseError"
  }
}

export class SerializeError extends Error {
  constructor(message: string) {
    super(message)

    this.name = "SerializeError"
  }
}
