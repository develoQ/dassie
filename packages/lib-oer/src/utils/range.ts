export type Range<T extends number | bigint> =
  | T
  | undefined
  | readonly [minimum: T | undefined, maximum: T | undefined]

export const parseNumberRange = (
  range: Range<number>
): { minimum: number | undefined; maximum: number | undefined } => {
  if (typeof range === "number") {
    return { minimum: range, maximum: range }
  }

  if (range == undefined) {
    return { minimum: undefined, maximum: undefined }
  }

  if (range[0] != undefined && range[1] != undefined && range[0] > range[1]) {
    throw new Error(`minimum value must be less than or equal to maximum value`)
  }

  return { minimum: range[0], maximum: range[1] }
}

export const parseBigintRange = (
  range: Range<bigint>
): { minimum: bigint | undefined; maximum: bigint | undefined } => {
  if (typeof range === "bigint") {
    return { minimum: range, maximum: range }
  }

  if (range == undefined) {
    return { minimum: undefined, maximum: undefined }
  }

  if (range[0] != undefined && range[1] != undefined && range[0] > range[1]) {
    throw new Error(`minimum value must be less than or equal to maximum value`)
  }

  return { minimum: range[0], maximum: range[1] }
}
