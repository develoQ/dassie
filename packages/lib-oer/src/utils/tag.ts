import { ParseError, SerializeError } from "./errors"
import { ParseContext, isSafeUnsignedInteger } from "./parse"

export const TAG_MARKER_UNIVERSAL = 0b00
export const TAG_MARKER_APPLICATION = 0b01
export const TAG_MARKER_CONTEXT = 0b10
export const TAG_MARKER_PRIVATE = 0b11

export const tagClassMarkerMap = {
  universal: TAG_MARKER_UNIVERSAL,
  application: TAG_MARKER_APPLICATION,
  context: TAG_MARKER_CONTEXT,
  private: TAG_MARKER_PRIVATE,
} as const

export const tagMarkerClassMap = [
  "universal",
  "application",
  "context",
  "private",
] as const

export type TagClass = keyof typeof tagClassMarkerMap
export type TagMarker = typeof tagClassMarkerMap[TagClass]

export const parseTag = (
  context: ParseContext,
  offset: number
):
  | ParseError
  | readonly [tag: number, tagClass: TagMarker, lengthOfTag: number] => {
  const { uint8Array } = context
  const tag = uint8Array[offset]

  if (tag == undefined) {
    return new ParseError(
      "unable to read tag - end of buffer",
      uint8Array,
      uint8Array.byteLength
    )
  }

  const tagClassMarker = (tag >>> 6) as TagMarker

  const tagFirstByte = tag & 0b0011_1111

  // When tag is less than 63, it is encoded in the lower six bits of the first byte
  if (tagFirstByte !== 0b0011_1111) {
    return [tagFirstByte, tagClassMarker, 1] as const
  }

  // When tag is 63 or greater, the first byte's lower six bits are set and the remaining bytes encode the tag in their lower seven bits until a byte with its upper bit set marks the end.
  let lengthOfTag = 1
  let tagValue = 0
  let nextByte: number | undefined
  do {
    nextByte = uint8Array[offset + lengthOfTag]

    if (nextByte == undefined) {
      return new ParseError(
        "unable to read tag - end of buffer",
        uint8Array,
        uint8Array.byteLength
      )
    }

    tagValue = (tagValue << 7) | (nextByte & 0b0111_1111)
    lengthOfTag++
  } while (nextByte & 0b1000_0000)

  return [tagValue, tagClassMarker, lengthOfTag] as const
}

export const serializeTag = (
  value: number,
  tagClass: TagMarker,
  uint8Array: Uint8Array,
  offset: number
) => {
  if (!isSafeUnsignedInteger(value)) {
    return new SerializeError(
      "unable to serialize tag - value is out of bounds"
    )
  }

  if (value < 0b0011_1111) {
    uint8Array[offset] = value | (tagClass << 6)
    return
  }

  uint8Array[offset] = (tagClass << 6) | 0b0011_1111

  let length = 0
  do {
    length++
  } while (value >= 1 << (7 * length))

  for (let index = length; index > 0; index--) {
    uint8Array[offset + index] =
      (value & 0b0111_1111) | (index === length ? 0 : 0b1000_0000)
    value = value >>> 7
  }
  return
}

export const predictTagLength = (value: number) => {
  if (!isSafeUnsignedInteger(value)) {
    return new SerializeError(
      "unable to serialize tag - value is out of bounds"
    )
  }

  if (value < 0b0011_1111) {
    return 1
  }

  let length = 0
  do {
    length++
  } while (value >= 1 << (7 * length))

  return length + 1
}
