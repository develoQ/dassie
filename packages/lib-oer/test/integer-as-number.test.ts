import { describe, test } from "vitest"

import {
  INT8_MAX_NUMBER,
  INT8_MIN_NUMBER,
  UINT8_MAX_NUMBER,
  UINT_MIN_NUMBER,
  int8Number,
  int16Number,
  int32Number,
  integerAsNumber,
  uint8Number,
  uint16Number,
  uint32Number,
} from "../src/integer-as-number"
import { hexToUint8Array } from "../src/utils/hex"
import { parsedOk, serializedOk } from "./utils/result"
import { getFixedLengthSamples } from "./utils/sample-numbers"

describe("integerAsNumber", () => {
  test("should be a function", ({ expect }) => {
    expect(integerAsNumber).toBeTypeOf("function")
  })

  describe("schema 0..uint8_max", () => {
    const schema = integerAsNumber([UINT_MIN_NUMBER, UINT8_MAX_NUMBER])

    test("should return an object", ({ expect }) => {
      expect(schema).toBeTypeOf("object")
    })

    test("should parse a zero", ({ expect }) => {
      const value = schema.parse(hexToUint8Array("00"))
      expect(value).toEqual(parsedOk(1, 0))
    })

    test("should serialize a zero", ({ expect }) => {
      const value = schema.serialize(0)
      expect(value).toEqual(serializedOk("00"))
    })

    test("should serialize a uint8", ({ expect }) => {
      const value = schema.serialize(0x12)
      expect(value).toEqual(serializedOk("12"))
    })
  })

  describe("schema int8_min..int8_max", () => {
    const schema = integerAsNumber([INT8_MIN_NUMBER, INT8_MAX_NUMBER])

    test("should serialize a negative int8", ({ expect }) => {
      const value = schema.serialize(-0x12)
      expect(value).toEqual(serializedOk("ee"))
    })
  })

  describe.each([
    ["uint8Number", uint8Number, false, 1],
    ["uint16Number", uint16Number, false, 2],
    ["uint32Number", uint32Number, false, 4],
    ["int8Number", int8Number, true, 1],
    ["int16Number", int16Number, true, 2],
    ["int32Number", int32Number, true, 4],
  ])("shortcut %s", (_, shortcutMethod, signed, byteLength) => {
    const schema = shortcutMethod()
    const sampleNumbers = getFixedLengthSamples(signed, byteLength)

    describe.each(sampleNumbers)("with sample number %s / %s", (value, hex) => {
      test("should parse the sample number", ({ expect }) => {
        const result = schema.parse(hexToUint8Array(hex))
        expect(result).toEqual(parsedOk(byteLength, Number(value)))
      })

      test("should serialize the sample number", ({ expect }) => {
        const result = schema.serialize(Number(value))
        expect(result).toEqual(serializedOk(hex))
      })
    })
  })
})
