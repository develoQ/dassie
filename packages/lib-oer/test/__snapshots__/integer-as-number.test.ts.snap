// Vitest Snapshot v1

exports[`integerAsNumber > schema uint8 with range 5..8 > should not parse a number that is too high 1`] = `
{
  "failure": [ParseError: unable to read fixed length integer of size 1 bytes - value 9 is greater than maximum value 8],
  "success": false,
}
`;

exports[`integerAsNumber > schema uint8 with range 5..8 > should not parse a number that is too low 1`] = `
{
  "failure": [ParseError: unable to read fixed length integer of size 1 bytes - value 1 is less than minimum value 5],
  "success": false,
}
`;

exports[`integerAsNumber > schema uint8 with range 5..8 > should not serialize a number that is too high 1`] = `
{
  "failure": [SerializeError: integer must be <= 8],
  "success": false,
}
`;

exports[`integerAsNumber > schema uint8 with range 5..8 > should not serialize a number that is too low 1`] = `
{
  "failure": [SerializeError: integer must be >= 5],
  "success": false,
}
`;
